Add additional structuring to Filter, Expression, and Literal

devinrsmith opened the pull request:
The interfaces of `io.deephaven.api.expression.Filter`, `io.deephaven.api.expression.Expression`, and `io.deephaven.api.value.Value` were originally developed and influenced based on the structure of the gRPC messages and engine capabilities; and they were explicitly called out due to lack of completeness with respect to these constructions in #829, #830, and #831.

Upon delving into SQL work in #3473, it became clear that it was time to update these interfaces with the more expressive structures that could better preserve the semantics and intentions of the SQL expressions. The alternative involves throwing up our hands and using `io.deephaven.api.RawString` everywhere instead.

For example, it was possible before to represent the filters of `Foo > Bar` and `Foo > 42`, but it was not possible to represent `Foo > Bar + 42` (again, without dropping down into `RawString`); and equivalently, an `Expression` could be `Bar`, or `42`, but not `Bar + 42` (essentially, limiting what could be represented in the case of view or update `Selectable`).

This PR greatly expands the expressiveness of these interfaces while also cleaning up the hierarchy. The main highlights:

* `Filter` is now an `Expression`
* `FilterCondition` renamed `FilterComparison`
* `Value` is gone; in its place, a more specific `Literal` structure
* Usages of `Value` replaced with the more generic `Expression` type (for example, LHS / RHS of `FilterComparison` are now `Expression`s instead of `Value`s)
* `ExpressionFunction` created to represent function calls without needing to drop into `RawString` (for example, we might want to have `io.deephaven.sql.Functions#sum_sql` as something that can be properly referenced).

This PR does not aim to update the gRPC layer at this time, mainly do to time considerations and that #3473 is limited to execution completely within the server environment. That said, we would likely benefit from updating the gRPC layer to be more expressive in the ways proposed here.

There's also potential in the future for the engine layer to benefit from the additional structuring here; potentially constructing more advanced filters and expressions without needing to compile strings.

devinrsmith commented:
In support of #3473

kosak commented on table-api/src/main/java/io/deephaven/api/filter/FilterOr.java:
```diff
@@ -41,6 +40,20 @@ public static FilterOr of(Iterable<? extends Filter> filters) {
     @Parameter
```
I wanted this comment to be on `FilterOr.of()` above (both overloads) but github won't let me click there.

It's tempting to do a little eager simplification based on the length of `filters`.  As in

```
if (filters.length == 0) {
   return Filter.ofFalse();  // empty OR is false
}
if (filters.length == 1) {
   return filters[0];
}
return builder().addFilters(filters).build();
```

This also reduces the number of cases where your `checkSize` method below would have to throw.

Again this would change the return value to just `Filter`

devinrsmith replied:
I think it makes more sense to do this one level "higher", at `io.deephaven.api.filter.Filter#or(java.util.Collection<? extends io.deephaven.api.filter.Filter>)`. Ditto for `and`.

devinrsmith replied:
I've update `Filter#or` and `Filter#and` in this style.

kosak commented on table-api/src/test/java/io/deephaven/api/filter/FilterComparisionTest.java:
```diff
@@ -0,0 +1,210 @@
+
+    @Test
+    void lhs() {
+        assertThat(FOO_EQ_42.lhs()).isEqualTo(FOO);
```
I feel somewhat motivated to consolidate the lhs, rhs, operation, and strings test into one helper method, as in

```
checkAllAspects(FOO_EQ_42, FOO, Operator.EQUALS, V42, "(Foo) == (42L)")
checkAllAspects(FOO_GT_42, FOO, Operator.GREATER_THAN, V42, "(Foo) > (42L)")
```

etc etc

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/SelectColumn.java:
```diff
@@ -191,23 +201,28 @@ public SelectColumn getOut() {
-        public void visit(Value rhs) {
-            rhs.walk((Value.Visitor) this);
+        public void visit(ColumnName rhs) {
+            out = new SourceColumn(rhs.name(), lhs.name());
```
I agree. I've used this pattern when I've introduced new interfaces. I was intending to _eventually_ do that for the QST types, but I guess now is as good as any? (Will add an additional non-trivial amount of diff, I think.)

devinrsmith replied:
I've updated Expression, Filter, and Literal visitors to use the return-type style.

devinrsmith commented on table-api/src/main/java/io/deephaven/api/filter/Filter.java:
```diff
@@ -15,8 +17,16 @@
+ * @see RawString
  */
-public interface Filter extends Serializable {
+public interface Filter extends Expression, Serializable {
```
So, an `Expression` is something that you can evaluate to _some_ value. `Foo + 32`, `Foo * my_method(Bar)`, `Foo > Bar` are expressions. A `Filter` is a specific expression that always evaluates to a _boolean_. So, `Foo > Bar`, `isNull(Foo)`, `Foo == Bar * 32` are filters.

Given that filtering is such an important concept for table operations, it has its own explicit type - ie, the interface is essentially `Table where(Collection<Filter> filters)` instead of `Table where(Collection<Expression> exprs)`.

kosak replied:
To clarify, I'm perfectly happy with `Filter` being its own type. What makes me nervous is `Filter` being an `Expression` rather than its own standalone type. It seems like you're saying that Filters are really always just expressions that have a boolean type. But it feels to me like some filters go beyond that.

Counterexample:

in SQL I can say

```
select * from Table where foo in (select something from somewhere)
```

but I can't substitute

```
select foo in (select something from somewhere) from Table
```

at least I don't think I can

The above (if you think it's valid) suggest to me that `Filter` is a different category of thing than Expression.

devinrsmith replied:
So, in Calcite everything is an "expression" (even what we are calling 'filters"), but they preserve more type information along the way (so, they know returns types, whereas DH is looser in this regards) and can error out when things don't match up. 

Regardless, we _need_ a way to represent assigning a boolean to a column: `table.view("Foo=MyValue>42")`. This is through the lens of `Selectable`, which is essentially a tuple of lhs `ColumnName` and rhs `Expression`. IMO, that means we have 3 options:

1. Have filter be an expression: `class Filter extends Expression`
2. Have filter be wrappable as an expression: `class ExpressionFilter extends Expression { Filter filter(); ...}`
3. Re-create the full structure of filter, but as mirrored onto expression: `ExpressionIsNull`, `ExpressionIsNotNull`, `ExpressionCondition`. The case could also be argued that when it comes to expressions, we should just fall back to the less-strongly-typed construct like `ExpressionFunction`, but I think that means we are depriving ourselves of the structure we already have w/ `Filter`.

My preference is 1 > 2 >> 3.

niloc132 replied:
I understand that calcite needs that information, but why does this non-sql api need that? Aren't we converting from SQL to QST, and not vice versa?

Or are you saying that, in context, we can't tell if calcite is going to assign an expression (which happens to be a filter) to a column? And, assuming that is true, can we not instead offer a Filter.toExpression() that will transform from a filter to an expression, rather than conflating the two?

devinrsmith replied:
It's not that we _need_ `Filter extends Expression`, I just think from a logical typing perspective it's the most natural (I see `Filter` as a specific type of `Expression`).

For example:

```java
table.select(Selectable.of(FOO, isNull(BAR)))
```
is how I imagine we'd get benefit out of having filter naturally be an expression. In the case where it's not and we have a wrapper or a type conversion, it would look like:

```java
table.select(Selectable.of(FOO, isNull(BAR).toExpression()))
```
or
```java
table.select(Selectable.of(FOO, ExpressionFilter.of(isNull(BAR))))
```

devinrsmith commented on table-api/src/main/java/io/deephaven/api/filter/FilterNot.java:
```diff
@@ -4,9 +4,12 @@
+
 /**
  * Evaluates to {@code true} when the given {@link #filter() filter} evaluates to {@code false}.
  */
```
There are reasons why I don't think its appropriate to do the simplifications at this level. There may be certain use cases where you want to preserve "nots" (up to a certain point) instead of "simplifying".

As a handwavy motivator, if we are translating the SQL expression (/filter) `NOT(NOT(NOT(Foo)))` into DHC, it's not _unreasonable_ to preserve that not logic all the way until it gets into the engine for execution instead of immediately simplifying to `NOT(Foo)`. Maybe we want to print out the QST and we want it to look as close as possible to the original users query as possible.

devinrsmith commented on table-api/src/main/java/io/deephaven/api/filter/FilterNot.java:
```diff
@@ -26,8 +29,23 @@ public static FilterNot of(Filter filter) {
+        return filter();
+    }
+
+    public final Optional<Filter> simplify() {
```
This simplify is specifically targeted at `FilterNot`. I've added javadoc and changed return value a bit.

devinrsmith replied:
Use a better name

devinrsmith commented on table-api/src/main/java/io/deephaven/api/filter/FilterIsNotNull.java:
```diff
@@ -3,32 +3,42 @@
 
     @Override
-    public final <V extends Visitor> V walk(V visitor) {
+    public final <V extends Filter.Visitor> V walk(V visitor) {
```
I think both patterns are supported, and it's more about how the visitor is implemented rather than the interface/name.

I typically like to structure my code so a visitor applies to a single node, and if recursion is needed, it will create a new visitor for each parent it needs to visit.

devinrsmith replied:
TBD naming to accept, acceptVisitor, dispatch?

niloc132 commented on java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java:
```diff
@@ -543,38 +557,68 @@ private static Reference reference(ColumnName columnName) {
-        public io.deephaven.proto.backplane.grpc.Value out() {
-            return Objects.requireNonNull(out);
+        @Override
+        public io.deephaven.proto.backplane.grpc.Value visit(int literal) {
```
I don't think these `Value` references need to be fully qualified any more.

niloc132 commented on table-api/src/main/java/io/deephaven/api/filter/FilterComparison.java:
```diff
@@ -5,60 +5,71 @@
          * @return the inverted operator
          */
-        public final Operator invert() {
+        public Operator inverse() {
```
isn't a verb better than a noun/adjective for a method name?

devinrsmith replied:
Oops, will do.

niloc132 commented on java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java:
```diff
@@ -596,62 +640,72 @@ private static CompareOperation adapt(Operator operator) {
-                            IsNullCondition.newBuilder().setReference(reference(isNull.column())).build())
+        public Condition visit(FilterIsNull isNull) {
+            if (!(isNull.expression() instanceof ColumnName)) {
+                // todo
```
needs a ticket

niloc132 commented on table-api/src/main/java/io/deephaven/api/filter/FilterIsNotNull.java:
```diff
@@ -3,33 +3,42 @@
-    public static FilterIsNotNull of(ColumnName column) {
-        return ImmutableFilterIsNotNull.of(column);
+    public static FilterIsNotNull of(Expression expression) {
+        return ImmutableFilterIsNotNull.of(expression);
```
Isn't this going to throw, since we only support checking non-null of a column name (`Only supports null checking a reference to a column`)?

devinrsmith replied:
From a logical perspective, we should be able to express a nullness check against an `Expression`. The limitations of the gRPC layer is called out in the PR description - I definitely think we should work on closing the gap, but it wasn't top-of-priority given that SQL is server-side (script session) only ATM.

devinrsmith replied:
#3609

niloc132 commented on table-api/src/main/java/io/deephaven/api/literal/LiteralInt.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Check
+    final void checkNotDeephavenNull() {
+        if (value() == Integer.MIN_VALUE) {
```
it would probably be better to get an dependency here on QueryConstants, or move those somewhere so that table-api can also reference them.

devinrsmith replied:
Yep - the dependency was why it wasn't already done. I'll look what it takes to move these constants to somewhere more accessible.

devinrsmith commented:
Expression if-then-else or switch, look into.

devinrsmith commented:
I spent a long time trying to reconcile potentially better structures around `Filter extends Expression`, extracting a different `BooleanExpression` interface, and various other machinations. I didn't feel any of them were particularly elegant.

In my mind, a lot of this comes down to "what does our structuring of inheritance 'mean'". For example, we have `ColumnName implements Selectable, Expression, Filter, Pair, JoinMatch, JoinAddition` - instead of thinking about this inheritance as "Column name is an Expression, ColumnName is a Filter, ..." it may be better to think about this as  "ColumnName can be used as an Expression, ColumnName can be used as a Filter, ...".

All of this is in favor of "ergonomics":

```java
ColumnName FOO = ColumnName.of("Foo");
ColumnName BAR = ColumnName.of("Bar");

// t1 = table.where("Bar"), BAR treated as Filter
Table t1 = table.where(BAR);

// t2 = table.view("Bar"), BAR treated as Selectable
Table t2 = table.view(BAR);

// t3 = table.view("Foo=Bar"), BAR treated as Expression
Table t3 = table.view(Selectable.of(FOO, BAR)); 
```

There might have been better names we could have used for these pattern:

`ColumnName implements ToSelectable, ToExpression, ToFilter, ...` and `Table where(Collection<? extends ToFilter> filters)`

That said, I think these sorts of naming changes would be superficial - I expect the visitor patterns and other methods on these interfaces would be exactly the same.

kosak commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilter.java:
```diff
@@ -211,161 +196,20 @@ public PreviousFilteringNotSupported(String message) {
-        }
+    @Override
+    default Filter invert() {
+        throw new UnsupportedOperationException("WhereFilters do not implement invert");
```
just a curiosity / skepticism issue, but is this thing really an Expression or Filter if it can't implement `invert`, `walk(Expression.Visitor)`, or `walk(Filter.Visitor)`

devinrsmith replied:
It's kind of a crappy hack; `WhereFilter` is not *really* a `io.deephaven.api.filter.Filter`, but decisions were made so that they would work in the context of `io.deephaven.engine.table.Table#where(Filter... filters)`.

kosak commented on engine/table/src/test/java/io/deephaven/engine/table/impl/select/WhereFilterTest.java:
```diff
@@ -16,60 +18,194 @@ public class WhereFilterTest extends TestCase {
+    }
+
+    public void testFooIsNotTrue() {
+        // This is *not* logically equivalent to "Foo in [false]"
```
Because you felt this was worth a comment, I suggest you finish out the comment with a clause or second sentence explaining why it's "*not* logically equivalent" so the reader doesn't have to pause to puzzle it out.

kosak reviewed (commented):
just some comments

devinrsmith commented:
Since the last review, I've updated this a bit - I've removed `FilterIsNotNull`; it's equivalent to `FilterNot<FilterIsNull>`, and better matches w/ the actual methods that DH exposes. I've added `FilterPattern`, `FilterMatches`, and `FilterQuick` to add some addition APIs that the web client UI would need in the future. I've added `LiteralString`

I'm not sure if `io.deephaven.api.expression.IfThenElse` is good idea right now - I'm thinking about removing it.

(Note: for some reason, new actions aren't being kicked off yet...)

devinrsmith commented on engine/table/src/main/java/io/deephaven/gui/table/filters/Condition.java:
```diff
@@ -8,24 +8,36 @@
  * classes support all types.
```
Hmmm - I didn't break this, but will see if there is a more appropriate class...

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/MatchFilter.java:
```diff
@@ -25,6 +27,16 @@ public class MatchFilter extends WhereFilterImpl {
     private static final long serialVersionUID = 1L;
 
+    public static MatchFilter ofStringValues(FilterMatches matches, boolean inverted) {
+        // Note: due to impl io.deephaven.engine.table.impl.select.MatchFilter.ColumnTypeConvertor.convertStringLiteral
```
Will do - https://github.com/deephaven/web-client-ui/issues/1243

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/ByteRangeFilter.java:
```diff
@@ -21,6 +21,22 @@
+
     final byte upper;
     final byte lower;
 
```
This code copies the literal logic that makeByteRangeFilter was using - I agree it's confusing, but I assume it was correct before, so continues to be correct. (If I were making more extensive changes, I would get rid of this flip-floppy constructor logic...)

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/ByteRangeFilter.java:
```diff
@@ -21,6 +21,22 @@
 
 public class ByteRangeFilter extends AbstractRangeFilter {
+    public static ByteRangeFilter lt(String columnName, byte x) {
+        return new ByteRangeFilter(columnName, QueryConstants.NULL_BYTE, x, true, false);
```
Literal copy of previous logic.

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/CharRangeFilter.java:
```diff
@@ -16,6 +16,22 @@
 
 public class CharRangeFilter extends AbstractRangeFilter {
+    public static CharRangeFilter lt(String columnName, char x) {
+        return new CharRangeFilter(columnName, QueryConstants.NULL_CHAR, x, true, false);
```
Ditto.

kosak replied:
I took a deeper look at this logic, and I now agree with you that it does the right thing. (I'm not in love with the prettiness of the logic, but that's a topic for another day)

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/SelectColumn.java:
```diff
@@ -178,36 +189,41 @@ default void validateSafeForRefresh(final BaseTable<?> sourceTable) {
-        public void visit(Value rhs) {
-            rhs.walk((Value.Visitor) this);
+        public SelectColumn visit(Filter rhs) {
+            return SelectColumnFactory.getExpression(String.format("%s=%s", lhs.name(), Strings.of(rhs)));
```
It's not actually identical - the `Strings.of(...)` calls are able to be more efficient when they are broken out like this since they have more explicit type information.

```java
        @Override
        public SelectColumn visit(Literal rhs) {
            return SelectColumnFactory.getExpression(String.format("%s=%s", lhs.name(), Strings.of(rhs)));
        }

        @Override
        public SelectColumn visit(Filter rhs) {
            return SelectColumnFactory.getExpression(String.format("%s=%s", lhs.name(), Strings.of(rhs)));
        }
        ...
```

`io.deephaven.api.Strings#of(io.deephaven.api.literal.Literal)` vs `io.deephaven.api.Strings#of(io.deephaven.api.filter.Filter)`

devinrsmith replied:
And we are already doing the visitor pattern for the optimized ColumnName case - I'll add a note in the code. This is a pattern I use elsewhere to, and will try and be proactive about putting this sort of note on them if you think it's valuable:

```java
        // Note: it might be tempting to consolidate all of the following getExpression calls to a common function, but
        // then we'd be losing the type information that allows us to call the more explicitly typed Strings#of(<type>)
        // methods.
```

kosak replied:
I think possibly you didn't notice my specific suggestion to keep the Strings.of part in my original suggestion.

devinrsmith replied:
AH! Yes, I can simplify it in that regard, will do.

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilter.java:
```diff
@@ -3,50 +3,35 @@
-
-    static WhereFilter ofInverted(Filter filter) {
-        return filter.walk(new Adapter(true)).getOut();
+        return (filter instanceof WhereFilter) ? (WhereFilter) filter : WhereFilterAdapter.of(filter);
```
So, the engine _is_ where we want to do these optimizations; as a client though constructing logic, we may or may not want to do that collapsing ourselves (and we currently don't).

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilter.java:
```diff
@@ -3,50 +3,35 @@
-
     static WhereFilter[] copyFrom(WhereFilter[] filters) {
         return Arrays.stream(filters).map(WhereFilter::copy).toArray(WhereFilter[]::new);
     }
```
I don't always follow this rule - but I'd prefer not to mess with unrelated code in this PR...

niloc132 commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java:
```diff
@@ -0,0 +1,421 @@
+                        return new MatchFilter(lhs.name(), rhs);
+                    case NOT_EQUALS:
+                        return new MatchFilter(MatchType.Inverted, lhs.name(), rhs);
+                    // Note: we can't assume IntRangeFilter - even though the rhs literal is an int, it might be against
```
This bothers me, since this method is only reachable when you're in a DH query engine of some kind. In fact, the only way that this method can be reached at all is from a WhereFilterAdapter instance, which plainly has a TableDefinition field. I don't immediately see a downside to dropping the static modifier on this inner-inner class (just have to move/rename the "public" static `of()` of the containing inner class), or pass the reference inward, so that this knowledge is available. Since the inner type FilterComparisonAdapter is private anyway, there can be no use case where, contrary to this comment, TableDefinition isn't available, so we should always know the appropriate type?

devinrsmith replied:
https://github.com/deephaven/deephaven-core/issues/3730

niloc132 commented on table-api/src/main/java/io/deephaven/api/Strings.java:
```diff
@@ -27,52 +31,85 @@ public static String of(ColumnName columnName) {
+    }
+
+    public static String of(FilterMatches matches, boolean encapsulate, boolean inverted) {
+        // <ColumnName> [icase] [not] in <value 1>, <value 2>, ... , <value n>
```
It _looks_ like this is reachable (and siblings too) by actually using filters as expressions? Isn't that in large part the point of the patch, that expressions can be expressed as filters?

Or will all such Filter types be guaranteed to be removed/rewritten by this point?

devinrsmith replied:
The Strings class is (unfortunately) dual-purpose right now. It is used both for descriptive purposes, but also for transforming QSTs into strings for the purpose of falling back to filter & expression parsing.

In the case of `FilterMatches`, when used in the context of a `where`, it will flow through `io.deephaven.engine.table.impl.select.WhereFilterAdapter` and construct the appropriate `io.deephaven.engine.table.impl.select.MatchFilter`; in the case where `FilterMatches` is used as part of an expression, then that isn't currently supported by the engine and will hit this "UnsupportedOperationException" branch.

I think there are 3 potential ways forward:

* Leave as-is
* Print a pretty string here anyways, and (expect to) fail during formula parsing
* Separate out Strings for descriptions, vs Strings for engine parsing as separate utilities 

In the short-term, I'm hoping to avoid the work involved with the last bullet point. I'll create a ticket to follow-up in this regards.

devinrsmith replied:
https://github.com/deephaven/deephaven-core/issues/3740

chipkent commented on table-api/src/main/java/io/deephaven/api/filter/README.md:
```diff
@@ -0,0 +1,96 @@
+
+# Foo == null || Foo.matches(...)
+include_match_include_null = t.where(not_(pattern(..., invert_pattern=True)))
+# include_match_include_null = t.where(or_([is_null("Foo"), pattern(...)]))
```
Remove this line?

chipkent commented on table-api/src/main/java/io/deephaven/api/filter/README.md:
```diff
@@ -0,0 +1,96 @@
+
+# Foo == null || !Foo.matches(...)
+exclude_match_include_null = t.where(not_(pattern(...)))
+# exclude_match_include_null = t.where(or_([is_null("Foo"), pattern(..., invert_pattern=True)]))
```
remove this line?

chipkent commented on table-api/src/main/java/io/deephaven/api/filter/README.md:
```diff
@@ -83,14 +85,35 @@ include_match_exclude_null = t.where(pattern(...))
+## Comparison operators
+
+Given all of the above, there is a design choice we have to make regarding comparison filters - does the inverse of a
+comparison filter results in another comparison filter?
```
```suggestion
comparison filter result in another comparison filter?
```

jmao-denver reviewed (commented):
The Python changes look good to me.

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterPatternImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+    private static final long serialVersionUID = 1L;
+
+    public static WhereFilterPatternImpl of(FilterPattern pattern, boolean inverted) {
```
We discussed that all of this inversion could be expressed as a filter that wraps the internal filter, and returns `selection.minus(wrapped.filter(...))`

devinrsmith replied:
Done

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereAllFilter.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * A Select filter that always returns the selection.
+ */
+public final class WhereAllFilter extends WhereFilterImpl {
```
We already have WhereNoneFilter, but I've gone ahead and deleted this and throwing an error where it was previously constructed.

rcaudy replied:
@niloc132 What is `WhereNoneFilter` for?

niloc132 replied:
In DHE, SelectNoneFilter is a singletone added so that the "search" filter works. Search allows the client to provide a value to search against all columns (that can be matched anyway) at once. 

in DHC, it is called WhereNoneFilter.

In both cases, `WhereFilterFactory.expandQuickFilter()` is given the table's definition and the search string - if no columns can be matched, then an empty array of filters will be returned, and rather than have everything match (as `Table.where()` will return everything...), the decision was made to create a singleton that just removed everything.

We probably could have instead created a new empty table from the same definition, but that wasnt what was implemented at the time. I can't speak further for the decisions made in that code, I wasn't directly involved, but there might have been a reason to do that.

rcaudy replied:
So, we "filter" the table to produce an empty result. Seems like a lot of work to get there. I get it, though.

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java:
```diff
@@ -0,0 +1,402 @@
+
+        @Override
+        public WhereFilter visit(Literal lhs) {
+            return original();
```
It's a case of multiple visitor implementations, and I don't see a good way to support delegation w/ multiple inheritance.

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java:
```diff
@@ -0,0 +1,402 @@
+        return MatchFilter.ofStringValues(matches, inverted);
+    }
+
+    public static WhereFilter of(ColumnName columnName, boolean inverted) {
```
I think `Filter.isTrue(Expression)` can fill in for cases where the user wants to filter based on ColumnName. I agree, it wasn't clear what the users / semantics of  `.where(columnName)`, `.where(columnName.invert())` was.

By removing it, there are 4 clear cases:

* `isTrue(Foo)` -> "!isNull(Foo) && Foo == true"
* `isFalse(Foo)` -> "!isNull(Foo) && Foo == false"
* `not(isTrue(Foo))` -> "isNull(Foo) || Foo == false"
* `not(isFalse(Foo))` -> "isNull(Foo) || Foo == true"

(Of course, this logic is expressed via MatchFilter impl and not ConditionFilter)

devinrsmith commented on table-api/src/main/java/io/deephaven/api/filter/FilterMatches.java:
```diff
@@ -0,0 +1,52 @@
+
+    public abstract Expression expression();
+
+    public abstract List<Literal> values();
```
Okay - I'll expand this type b/c I agree. That said, there is room for a `QueryScope extends Expression` type in the future that we don't currently have.

rcaudy replied:
You need to support `FilterMatches` in some way, but it could be via equals, in, and case-insensitive String-specific variants of each (4 total).
I still think you need param support.

devinrsmith replied:
I think we should follow-up later w/ https://github.com/deephaven/deephaven-core/issues/3784

devinrsmith replied:
And https://github.com/deephaven/deephaven-core/issues/3785

devinrsmith commented on table-api/src/main/java/io/deephaven/api/filter/FilterMatches.java:
```diff
@@ -0,0 +1,52 @@
+        return ImmutableFilterMatches.builder();
+    }
+
+    public abstract Expression expression();
```
We're trying to be as generic enough to capture the semantics that one could reasonably want to express (or express via SQL), even though our engine might not support it for now.

As a quasi example in our parser-like syntax:

```
Foo + Bar in "foobar"
```

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/MatchFilter.java:
```diff
@@ -25,6 +28,17 @@ public class MatchFilter extends WhereFilterImpl {
+            throw new IllegalArgumentException("MatchFilter only supports filtering against a column name");
+        }
+        return new MatchFilter(
+                matches.caseInsensitive() ? CaseSensitivity.IgnoreCase : CaseSensitivity.MatchCase,
```
I think there is good reason to add string-specific filters. In fact, I've done that in a different PR exposing fast impls for `starts_with`, `ends_with`, `contains`, and case-insensitive versions of all of those.

It's is funky to have case-sensitivity options for non-string columns.

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterPatternImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+    @Override
+    public String toString() {
+        return "WhereFilterPatternImpl{" +
```
I mean, I'm not sure of a good way to make it pretty...

devinrsmith replied:
I've made a couple of WhereFilter impl toString prettifications.

devinrsmith commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterPatternImpl.java:
```diff
@@ -0,0 +1,181 @@
+        private final Predicate<T> predicate;
+
+        public ObjectChunkFilterPredicate(Predicate<T> predicate) {
+            this.predicate = Objects.requireNonNull(predicate);
```
The alternative in this current case (modulo some invert talks below) would be to double the number of chunk filters again (from 4 to 8). In general, I don't like the abstractions / implementations we currently have around `io.deephaven.engine.table.impl.chunkfilter.ChunkFilter` b/c all of the implementations essentially end up repeating exactly the same logic.

In my testing, I found no difference between the two implementations (not in 11 nor 17).

devinrsmith replied:
I've expanded this to 8 separate ObjectChunkFilters (instead of a composable predicate plumbed through single ObjectChunkFilter).

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java:
```diff
@@ -32,27 +32,90 @@ public static WhereFilter of(FilterNot<?> not) {
+    }
+
+    public static WhereFilter of(Filter filter, boolean inverted) {
+        return inverted ? filter.walk(new WhereFilterAdapter(true)) : of(filter);
```
Why not just pass `inverted` down?

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java:
```diff
@@ -32,27 +32,90 @@ public static WhereFilter of(FilterNot<?> not) {
+    }
+
+    public static WhereFilter of(FilterNot<?> not, boolean inverted) {
+        return inverted ? of(not.invert()) : of(not);
```
You said would simplify this one, too.

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java:
```diff
@@ -32,27 +32,90 @@ public static WhereFilter of(FilterNot<?> not) {
+        return inverted ? of(not.invert()) : of(not);
+    }
+
+    public static WhereFilter of(FilterOr ors, boolean inverted) {
+        // !A && !B && ... && !Z (inverted)
+        // A || B || ... || Z (regular)
+        return inverted ? of(ors.invert(), false) : of(ors);
+    }
+
+    public static WhereFilter of(FilterAnd ands, boolean inverted) {
+        // !A || !B || ... || !Z (inverted)
+        // A && B && ... && Z (regular)
+        return inverted ? of(ands.invert(), false) : of(ands);
+    }
```
You plan to eliminate extraneous falses.

deephaven-internal commented:
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/2563
Conceptual: https://github.com/deephaven/deephaven.io/issues/2562
Reference: https://github.com/deephaven/deephaven.io/issues/2565
Blog: https://github.com/deephaven/deephaven.io/issues/2564
